packages/react-query/src/index.ts

```diff
@@ -209,13 +209,52 @@ export type QueryClient<TApiRouter extends FlatApiRouter> =
       }
     : never
 
+function isEmptyObject(obj: any): boolean {
+  return (
+    obj != null && typeof obj === 'object' && !Array.isArray(obj) && Object.keys(obj).length === 0
+  )
+}
+
+function sortObjectDeep(obj: any): any {
+  if (obj == null || typeof obj !== 'object' || Array.isArray(obj)) {
+    return obj
+  }
+
+  return Object.keys(obj)
+    .sort()
+    .reduce((result, key) => {
+      result[key] = sortObjectDeep(obj[key])
+      return result
+    }, {} as any)
+}
+
+function normalizePayload(payload?: any): any {
+  if (!payload) return undefined
+
+  const normalized = {
+    pathParams: payload.pathParams,
+    query: payload.query,
+    headers: payload.headers,
+  }
+
+  const filtered = Object.entries(normalized).reduce((acc, [key, value]) => {
+    if (value != null && !isEmptyObject(value)) {
+      acc[key] = sortObjectDeep(value)
+    }
+    return acc
+  }, {} as any)
+
+  return Object.keys(filtered).length > 0 ? filtered : undefined
+}
+
 export function queryKey(method: string, path: string | number | symbol, payload?: any) {
-  const payloadKey = {
-    pathParams: payload?.pathParams ?? {},
-    query: payload?.query ?? {},
-    headers: payload?.headers ?? {},
+  const normalizedPayload = normalizePayload(payload)
```

Comment:
Please omit headers from payload. 

The key should be like this

```
["GET", "/path", { query: { ... }, pathParams: { ... }}, headers]
```

Since some does not care headers but some care

If we invalidate key 

```
["GET", "/path", { query: { ... }, pathParams: { ... }}]
```

The following key will also be invalidated 

```tsx
["GET", "/path", { query: { ... }, pathParams: { ... }}, headers]
```